packages/util/ratio.go

```diff
@@ -107,7 +107,7 @@ func (ratio *Ratio32) Read(r io.Reader) error {
 	rr := rwutil.NewReader(r)
 	ratio.A = rr.ReadUint32()
 	ratio.B = rr.ReadUint32()
-	if rr.Err == nil && ratio.HasZeroComponent() {
+	if rr.Err == nil && !(ratio == &Ratio32{} || !ratio.HasZeroComponent()) {
```

Comment:
@dessaya here it has a cyclic importing problem. And we need to support `0:0` here too. Therefore, I have to do in this way. Is it ok?

Reply:
Well... I don't know. Again, this validation is specific to gas. From the point of view of `Ratio32`, 0:0 is invalid (since mathematically 0/0 is undefined). And both 0:x or x:0 are "dangerous" because they might panic withzero division error.

But OK, maybe the simplest solution is to add the concept of a "Zero" Ratio32, so the invariant of Ratio32 can be:

> Either A != 0 and B != 0, or both A == 0 and B == 0.

Then:

```
func (ratio Ratio32) IsZero() bool {
	return ratio.A == 0 && ratio.B == 0
}

func (ratio Ratio32) IsValid() bool {
	return ratio.IsZero() || !ratio.HasZeroComponent()
}
```

Then:

```suggestion
	if rr.Err == nil && ratio.IsValid() {
```

Or just ignore all my comments and do what you thik is best. This is just me nitpicking... :stuck_out_tongue: